### PR TITLE
Atomic metadata operations! (master)

### PR DESCRIPTION
**Pull request by korydraughn**
This PR has dependencies on libraries that are not part of irods-externals.

Dependencies:
- [nanodbc](https://github.com/nanodbc/nanodbc) library
- [fmtlib](https://github.com/fmtlib/fmt) library

### CODE REVIEW

File: unit_tests/src/test_atomic_apply_metadata_operations.cpp [hunk 1]
```diff
@@ -0,0 +1,477 @@
...
+                    {"attribute", "this_was_set_atomically"},
+                    {"value", "it worked!"}
+                }
+            })}
+        }.dump();
+
+        char* json_error_string{};
+
+        REQUIRE(rc_atomic_apply_metadata_operations(&_conn, json_input.c_str(), &json_error_string) == 0);
+        REQUIRE(json_error_string == "{}"s);
```
Comment (trel) [inline unit_tests/src/test_atomic_apply_metadata_operations.cpp]: Codacy Issue found: [String literal compared with variable 'json_error_string'. Did you intend to use strcmp() instead?](https://app.codacy.com/app/trel/irods/pullRequest?prid=4460794)

Comment (korydraughn) [reply unit_tests/src/test_atomic_apply_metadata_operations.cpp]: The `""s` is a `std::string`. The trailing `s` on the empty string literal is the key here.

Comment (korydraughn) [reply unit_tests/src/test_atomic_apply_metadata_operations.cpp]: Codacy doesn't understand the trailing `s`.

Comment (trel) [inline unit_tests/src/test_atomic_apply_metadata_operations.cpp]: Codacy Issue found: [String literal compared with variable 'json_error_string'. Did you intend to use strcmp() instead?](https://app.codacy.com/app/trel/irods/pullRequest?prid=4460794)

Comment (trel) [inline unit_tests/src/test_atomic_apply_metadata_operations.cpp]: Codacy Issue found: [String literal compared with variable 'json_error_string'. Did you intend to use strcmp() instead?](https://app.codacy.com/app/trel/irods/pullRequest?prid=4460794)

Comment (trel) [inline unit_tests/src/test_atomic_apply_metadata_operations.cpp]: Codacy Issue found: [String literal compared with variable 'json_error_string'. Did you intend to use strcmp() instead?](https://app.codacy.com/app/trel/irods/pullRequest?prid=4460794)

Comment (trel) [inline unit_tests/src/test_atomic_apply_metadata_operations.cpp]: Codacy Issue found: [String literal compared with variable 'json_error_string'. Did you intend to use strcmp() instead?](https://app.codacy.com/app/trel/irods/pullRequest?prid=4460794)

File: plugins/api/src/atomic_apply_metadata_operations.cpp [hunk 2]
```diff
@@ -0,0 +1,583 @@
...
+    }
+
+    auto to_bytes_buffer(const std::string& _s) -> bytesBuf_t*
+    {
+        const auto buf_size = _s.length() + 1;
+
+        char* buf = new char[buf_size]{};
+        std::strncpy(buf, _s.c_str(), _s.length());
+
+        bytesBuf_t* bbp = new bytesBuf_t{};
```
Comment (trel) [inline plugins/api/src/atomic_apply_metadata_operations.cpp]: Codacy Issue found: [Uninitialized variable: bbp](https://app.codacy.com/manual/trel/irods/pullRequest?prid=4460794)

File: unit_tests/src/test_atomic_apply_metadata_operations.cpp (hunk 1 above)
Comment (trel) [inline unit_tests/src/test_atomic_apply_metadata_operations.cpp]: Codacy Issue found: [String literal compared with variable 'json_error_string'. Did you intend to use strcmp() instead?](https://app.codacy.com/manual/trel/irods/pullRequest?prid=4460794)

File: plugins/api/src/atomic_apply_metadata_operations.cpp (hunk 2 above)
Comment (trel) [inline plugins/api/src/atomic_apply_metadata_operations.cpp]: Codacy Issue found: [Uninitialized variable: bbp](https://app.codacy.com/manual/trel/irods/pullRequest?prid=4460794)

File: unit_tests/src/test_atomic_apply_metadata_operations.cpp (hunk 1 above)
Comment (trel) [inline unit_tests/src/test_atomic_apply_metadata_operations.cpp]: Codacy Issue found: [String literal compared with variable 'json_error_string'. Did you intend to use strcmp() instead?](https://app.codacy.com/manual/trel/irods/pullRequest?prid=4460794)

**Pull request closed without merging**